docs/blueprints/EP038.rst

```diff
@@ -0,0 +1,169 @@
+and values being the rate to pace them to.
+The pacing value is a string formatted as
+``{max_concurrent}/{time_period}{time_unit}``,
+where ``max_concurrent`` is the max amount of times
+the action is allowed to be executed
+within the ``time_period`` using the given ``time_unit``.
+Here's an example:
+
+.. code-block:: python
+
+    action_paces = {
+        "send_flow_mod": "50/2seconds",
+    }
+
+How the ``time_period`` is interpreted is controlled by the pacing strategy.
+Pacing strategy can be set using the ``pacing_strategy`` option in ``kytos.conf``.
+Available strategies include:
+
+ - `fixed_window` - Actions are paced in a fixed window,
+   where actions are paced in fixed blocks of time
+   that reset after the amount of time in the refresh period has passed.
+ - `elastic_window` - Actions are paced in an elastic window,
+   which acts like a fixed window, except if the pace is exceeded
+   then the window refresh period is doubled.
+ - `moving_window` - Actions are paced in a moving window,
+   where actions are paced in a dynamic block of time
+   that trails behind the current time.
+
+
+
+Pacing an Action
+================
+
+In order to add pacing functionality to an action,
+the pacing controller must be integrate into the NApp.
+To do so, add in a ``PacerNappWrapper`` to wrap
+around the pacer controller provided by the Kytos Controller.
+Then with the wrapped pacer, inject the configuration from the
+NApp's settings. Here's an example of this:
+
+.. code-block:: python
+
+    from kytos.core.pacing import PacerNappWrapper
+    from napps.kytos.my_napp import settings
+
+    class Main(KytosNApp):
+        ...
+
+        def setup(self):
+            ...
+
+            self.pacer = PacerNappWrapper(
+                "mynapp",
+                self.controller.pacer
+            )
+
+            self.pacer.inject_config(
+                settings.action_paces
+            )
+
+            ...
+
+With the pacer added to the NApp at startup,
+it can then be used at runtime.
+To use the pacer, use the ``hit`` or ``ahit`` methods.
+The ``hit`` and ``ahit`` methods take in the name of an action,
+and a set of optional keys, which will then
+pause execution of the current thread or task until
+the action is allowed to be executed.
+
+If we where to want to pace an action basedon a the given switch,
+we would use the switch's ID, as one of the keys.
+Here's an example of such an implementation:
+
+.. code-block:: python
+
+    class Main(KytosNApp):
+        ...
+
+        def send_switch_event(self, switch: Switch):
+            """Send an event about the switch."""
+            self.pacer.hit("mynapp.switch_event", switch.dpid)
+            self.controller.buffers.app.put(
+                ...
+            )
+
+        async def asend_switch_event(self, switch: Switch):
+            """Asynchronously send an event about the switch."""
+            await self.controller.pacer.ahit("mynapp.switch_event", switch.dpid)
```

Comment:
Did you mean `self.pacer.ahit(*args)` or indeed `self.controller.pacer.ahit(*args)`? 

Also, looks like `PacerNappWrapper` exposing both `ahit` and `hit` might still be using a janus queue? Last we spoke it's was desired not to compound on extra janus queue to expose this in the same abstraction. Let me know how that unfolded or if there new discoveries.

Reply:
No, janus queue isn't used here. And you are correct its meant to be `self.pacer.ahit`.

Reply:
Got it.